src/rust/engine/src/nodes.rs

```diff
@@ -166,37 +168,62 @@ impl WrappedNode for Select {
           task: task.clone(),
           entry: Arc::new(self.entry.clone()),
         }),
-        &tasks::Rule::Intrinsic(Intrinsic { product, input })
-          if product == context.core.types.snapshot && input == context.core.types.path_globs =>
+	&Rule::Intrinsic(Intrinsic { product, input })
+	  if product == types.directory_digest && input == types.input_file_content =>
+          {
+            let context = context.clone();
+            let store = context.core.store();
+
+            self.select_product(&context, types.input_file_content, "intrinsic")
+              .and_then(move |files_content: Value| {
+                let filename = externs::project_str(&files_content, "path");
+                let bytes: bytes::Bytes = externs::project_bytes(&files_content, "content").into();
+
+                let digest = store.store_file_bytes(bytes, false)
+                  .map_err(|e| throw(&e));
+
+                digest.and_then(move |digest: hashing::Digest| {
+                  let mut directory = bazel_protos::remote_execution::Directory::new();
+                  directory.mut_files().push({
+                    let mut file_node = bazel_protos::remote_execution::FileNode::new();
+                    file_node.set_name(filename);
+                    file_node.set_digest((&digest).into());
+                    file_node
+                  });
+
+                  store.record_directory(&directory, true)
+                    .map_err(|e| throw(&e))
+                    .map(move |digest: hashing::Digest| Snapshot::store_directory(&context.core, &digest))
+                })
+              })
+            .to_boxed()
+	},
+        &Rule::Intrinsic(Intrinsic { product, input })
+          if product == types.snapshot && input == types.path_globs =>
         {
           let context = context.clone();
           let core = context.core.clone();
           self
-            .select_product(&context, context.core.types.path_globs, "intrinsic")
+            .select_product(&context, types.path_globs, "intrinsic")
```

Comment:
It tends to make it harder to review with these sorts of changes, *especially* with rustfmt involved. While I strongly agree with the spirit, I found it difficult when reviewing the previous commits of this you'd shown me to know which lines you'd really changed versus which ones were cosmetic. I *think* these could stay since I've maybe already reviewed them a couple times, but in general I try to avoid making these sorts of abbreviations personally.